File: 1972-rotating-the-box/rotating-the-box.java
class Solution {
    public char[][] rotateTheBox(char[][] box) {
        int rows = box.length;
        int cols = box[0].length;
        char[][] result = new char[cols][rows];

        for (int i = 0; i < rows; i++) {
            int emptySpot = cols - 1;
            for (int j = cols - 1; j >= 0; j--) {
                if (box[i][j] == '#') {
                    box[i][j] = '.';
                    box[i][emptySpot--] = '#';
                } else if (box[i][j] == '*') {
                    emptySpot = j - 1;
                }
            }
            for (int j = 0; j < cols; j++) {
                result[j][rows - 1 - i] = box[i][j];
            }
        }

        return result;
    }
}

File: 1972-rotating-the-box/rotating-the-box.rs
impl Solution {
    pub fn rotate_the_box(r#box: Vec<Vec<char>>) -> Vec<Vec<char>> {
        let rows = r#box.len();
        let cols = r#box[0].len();
        let mut result = vec![vec!['.'; rows]; cols];

        for i in 0..rows {
            let mut empty_spot = cols - 1;
            for j in (0..cols).rev() {
                if r#box[i][j] == '#' {
                    result[empty_spot][rows - 1 - i] = '#';
                    empty_spot -= 1;
                } else if r#box[i][j] == '*' {
                    result[j][rows - 1 - i] = '*';
                    empty_spot = j - 1;
                }
            }
        }

        result
    }
}